inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/FlinkTubeMQConsumer.java

```diff
@@ -0,0 +1,355 @@
+                            consumeResult.getErrMsg());
+                }
+
+                Duration idleTime = Duration.between(lastConsumeInstant, Instant.now());
+                if (idleTime.compareTo(maxIdleTime) > 0) {
+                    ctx.markAsTemporarilyIdle();
+                }
+
+                continue;
+            }
+
+            List<Message> messageList = consumeResult.getMessageList();
+            lastConsumeInstant = Instant.now();
+
+            List<T> records = new ArrayList<>();
+            lastConsumeInstant = getRecords(lastConsumeInstant, messageList, records);
+
+            synchronized (ctx.getCheckpointLock()) {
+
+                for (T record : records) {
+                    ctx.collect(record);
+                }
+                currentOffsets.put(
+                        consumeResult.getPartitionKey(),
+                        consumeResult.getCurrOffset());
+            }
+
+            ConsumerResult confirmResult = messagePullConsumer
+                    .confirmConsume(consumeResult.getConfirmContext(), true);
+            if (!confirmResult.isSuccess()) {
+                if (!(confirmResult.getErrCode() == TErrCodeConstants.BAD_REQUEST
+                        || confirmResult.getErrCode() == TErrCodeConstants.NOT_FOUND
+                        || confirmResult.getErrCode() == TErrCodeConstants.ALL_PARTITION_FROZEN
+                        || confirmResult.getErrCode() == TErrCodeConstants.NO_PARTITION_ASSIGNED
+                        || confirmResult.getErrCode() == TErrCodeConstants.ALL_PARTITION_WAITING
+                        || confirmResult.getErrCode() == TErrCodeConstants.ALL_PARTITION_INUSE)) {
+                    LOG.warn("Could not confirm messages to tubemq (errcode: {},errmsg: {}).",
+                            confirmResult.getErrCode(),
+                            confirmResult.getErrMsg());
+                }
+            }
+        }
+    }
+
+    private Instant getRecords(Instant lastConsumeInstant, List<Message> messageList, List<T> records)
+            throws Exception {
+        if (messageList != null) {
+            lastConsumeInstant = Instant.now();
+            if (!innerFormat) {
+                for (Message message : messageList) {
+                    T record = deserializationSchema.deserialize(message.getData());
+                    records.add(record);
+                }
+            } else {
+                List<RowData> rowDataList = new ArrayList<>();
+                ListCollector<RowData> out = new ListCollector<>(rowDataList);
+                for (Message message : messageList) {
+                    deserializationSchema.deserialize(message.getData(), (Collector<T>) out);
+                }
+                rowDataList.forEach(data -> records.add((T) data));
+            }
+        }
+        if (CollectionUtils.isEmpty(messageList)) {
+            return lastConsumeInstant;
+        }
+
+        // other logic
```

Comment:
don't think the code will be run here, collectionUtils does null verification already